### ISSUE TITLE
Problems with official examples

### ISSUE THREAD

**Issue by jmaye** [tensorflow/models#2489]
I'm playing with Tensorflow for a week now and I was able to hook up another database to the models in the official folder (https://github.com/tensorflow/models/blob/master/official/resnet/imagenet_main.py). I like this high-level API much and everything seems to work fine apart the following issues:

- when using a large database containing millions of images, the workstation just explodes. High RAM usage, events.out.tfevent on the order of several gigabytes. I realized that this problem is related to the fact that each time a checkpoint is saved or a summary, the entire list of images is saved to disk:) I could circumvent this problem by using placeholders for filenames and variables, feeding them using a hook on the train function. I don't know if this the recommended way. I've read somewhere that one might want to transfer the database into TFRecords. However, what's best there: a single file or multiple files?

- I had to add something like logits += tf.constant(1e-8) to avoid NaN at some point

- training proceeds nicely apart that something is destroyed each time a checkpoint is restored (after stop and restart or when the train function finishes in the cycle while loop).

![screenshot_2017-10-04_09-40-25](https://user-images.githubusercontent.com/2109115/31165885-04564040-a8ed-11e7-88e8-a9cf58070942.png)

Is this something expected? I suspect that some parameters are not well restored (batch norm moving mean or moving variance?). If I run the example on a smaller database with fewer classes, this does not seem that dramatic. But I guess it is because I haven't run it for long. Maybe the moving mean/variance is really well estimated, then destroyed, causing those curves.

I've posted this question at several places. Also, I've seen people having the same issue in different places. However, I found no convincing answer. Thanks for your help.

**Comment by jmaye** [comment]
The problem might be more subtle, a potential misuse of the function one_hot. Indeed, I have integer labels that are not sequential, e.g. labels = [0, 10, 20, 40, 50], then I was using it like:

onehot_labels = tf.one_hot(labels, 5)

I just realized that the label 10, 20, 40, 50 are all equal in one_hot. I was assuming that internally there was a dictionary doing it correctly, but apparently you need to have sequential labels.

I just launched a new experiment by doing the dictionary myself.

**Comment by jmaye** [comment]
That label stuff did not solve the problem, there is indeed a problem when restoring the model. I've tried everything I could, no way:(

**Comment by reedwm** [comment]
@jmaye you are correct `one_hot` requires sequential labels. Make sure all the indices to `one_hot` are less than the second argument to `one_hot`.

What do you mean when you say you hooked up another database to the model? Does that mean you no longer use the Imagenet data in TFRecords format? Having more images should not increase RAM usage significantly, so perhaps the increased RAM usage is due to database you're using. It is not recommended to use placeholders for filenames and variables, and doing so should not use significantly less RAM.

Does the issue with restoring checkpoints occur if you use the Imagenet TFRecords data without any modifications to the code?

**Comment by jmaye** [comment]
Thanks for you reply!

What I mean is the following, I'm adapting the file imagenet_main.py or cifar10_main.py so that I can use them to train a network on another database than imagenet or cifar10.

Concretely, this boils down to doing something like:

dataset = tf.contrib.data.Dataset.from_tensor_slices((filenames, labels)),

where filenames and labels are numpy arrays containing respectively the filepaths of the images and the integer labels. The array filenames and labels contain millions of elements.

Doing this, I got huge log files (events.out.tfevents* and checkpoints of several GB). Inspecting those files, it turns out that the filenames are indeed written to the log files repeatedly.

I found a workaround to this by googling a bit. It consists in using placeholders instead of the actual numpy arrays. The real filenames and labels are fed via a hook to the train function (overriding after_create_session function to feed dict).

The problem now is, as mentioned, that every time I restore a checkpoint, the loss increases and the training accuracy decreases. I have to wait nearly an epoch (or two) to retrieve the old values. However it seems that it's not totally starting from scratch. My second workaround to this, I avoid restoring checkpoints and do the learning infinitely. It seems to work good for now. My only fear is that when I will be reusing the network for prediction, I won't get the same accuracy:)

Two things come to my mind concerning this issue:
- something is not well restored (batch normalization moving means or variances for instance?)
- the dataset input pipeline works only with TFRecords

I know that you will say I shall look on stackoverflow since it's not the original code. I'm not lazy, I did. I even found some people having exactly the same issue I have. Unfortunately, there is no single convincing answer to these threads.

If I were to use TFRecords, what would be the strategy? One big file for all the data or cut in several smaller files?

I will try anyway to reproduce the restore behavior on imagenet or cifar10 example.

**Comment by jmaye** [comment]
Just finished learning after more than 30 epochs, training accuracy was 95%. Restarted the stuff in evaluation mode and it drops to 10%, hence unusable. Don't know what's wrong with this, is it related to this issue: https://github.com/tensorflow/tensorflow/issues/1122

**Comment by jmaye** [comment]
I just did some more testing with the cifar10_main.py example and noticed the following:

- reported training accuracy is about 73% at 4000 steps
- model is saved and restored
- training accuracy starts at 88%

I have the same behavior when I reduce the number of classes of my problem.

Am I missing something? Is this behavior normal?

To complete the story, I notice that the cross-entropy is not jumping.

**Comment by jmaye** [comment]
This last issue is now solved. Actually, tf.metrics.accuracy maintains a running statistics. This was a bit counter-intuitive to me:) This explains the jump, because when you restart you forgot the first iterations where the accuracy was lower, and hence you go higher.

**Comment by jmaye** [comment]
I'm now doing something like:

    batch_accuracy =\
        tf.reduce_mean(tf.to_float(tf.equal(predictions['classes'],
                       tf.argmax(onehot_labels, axis=1))))

**Comment by reedwm** [comment]
Thank you for looking into this. To clarify: You said earlier that "every time I restore a checkpoint, the loss increases and the training accuracy decreases". But now it seems the accuracy increases when you restore, due to tf.metrics as you stated. In what cases did the accuracy decrease?

Also, when I say to go to StackOverflow, I mean to post a question, not just look at existing questions and answers. For custom code, StackOverflow is usually more useful than Github Issues, unless you narrowed down the issue to a specific chunk of code.

Do you get large checkpoints and summary files when running without modifications to the code? I'm not sure why filenames are being written to disk, or why the placeholder strategy would help.

**Comment by jmaye** [comment]
I understand my comments are a little bit confusing. Here is a summary of the different things I've done:
- based on imagenet_main.py, I did a similar database loading mechanism, i.e.:
  dataset = tf.contrib.data.Dataset.from_tensor_slices((filenames, labels)),
  where filenames and labels are huge python string/integer lists (4M images/labels)
  Problem: huge log files! events.out.tfevents in the order of several Giga and continuously growing for 
  instance.
  It turns out, by inspecting those files, that the entire list of filenames and labels is being written to it, and 
  this several times (indeed, each time a summary is written).
  Workaround: use placeholders instead of python string lists, hook up some feeding mechanism to feed 
                        the actual filenames and labels at session start
  Comment: I think this is not the way to go, but I would be glad to hear or see an example of managing a 
  large database with the Dataset pipeline
- I discovered on two databases (cifar10 and my own small database of 250 classes) that the training 
  accuracy jumps up after restart. This was a bit counter-intuitive to me. It turns out that the metrics used 
  in the official example actually keeps an history of the accuracy. To me, it makes more sense to monitor 
  what's happening in the current batch I'm processing and it is also related to the current loss.
- I'm still stuck on the last remaining issue here. My large database of 4M images got a nice training after 
  300K iterations, about 95% training accuracy reported. Using this for testing on the SAME database or 
  plainly restarting the training got me down to a training accuracy of 10%! I have no clue where the 
  problem originates. Might be a problem of imbalance in the dataset (some classes have lots of images, 
  some less), or a problem of randomization (buffer size not well chosen). Anyway, it would nice to select 
  batches differently, but I don't see how to do that in Dataset pipeline. My idea is to create a batch of 
  data by selecting N random images from M random classes, then all batches are nicely balanced.

If things go on like this, I think I will just cancel my attempt on this Dataset pipeline or even on Tensorflow. It's kind of surprising that I cannot get something after nearly 2 weeks of trials when this framework was so much praised. Not to say that I could have results on other frameworks after a couple of hours:) But apparently, Tensorflow is supposed to be more flexible and tunable.

**Comment by reedwm** [comment]
For the first point about huge log files, I'm not sure where those are coming from. The input pipeline with `tf.contrib.data` should not be adding any summaries. @nealwu can you confirm this?

For the third point about low train accuracy, its impossible to debug without any code. Posting the 4M images is infeasible, but can you post the code to see what is happening? Posting the code to StackOverflow would also probably get you better help.

**Comment by jmaye** [comment]
I agree this is pretty hard without seeing the code:) I will see what I can do to post it.
Anyway, here is the status of my experiments today:
- I limited the number of classes to 1000, then 2000, then 5000, then 7500. Everything works as expected until 7500, which seems to be the breaking point. When I say everything works fine, I can stop the training and restart it to get the same loss and training accuracy. Having more classes lead to starting with accuracy of zero but rapidly going to the previous accuracy. It's like if something is lost somewhere. Debugging more...

**Comment by jmaye** [comment]
I think I might use another method for creating the dataset than from_tensor_slices, I've seen the list_files, from which you give a wild card.

**Comment by jmaye** [comment]
For the problem of loss, could it be a problem of imbalance, randomization? Ideally, I would love to pick let's say 30 classes at random out of the 70k available, then 20 images per classes to create a batch.

**Comment by jmaye** [comment]
New insights on the issue! I changed everything in my input pipeline using now TextLineDataset and other stuff. I'm able to have an insignificant initial loading time (whereas I was waiting 5 minutes before). And best of all, I can balance my batches the way I want, let's say 15 samples from 20 classes for instance.

What I discover:
- probably for reproducibility, the random seed was set to 1 by default. Hence, when you use repeat on your dataset, the data samples are exactly in the same order. I changed that to None, working better.
- with randomness set appropriately, the loss takes way more time to decrease
- I can now stop learning and restore a checkpoint, I have exactly the same loss as when I excited and the same accuracy

What happened before:
- the optimizer was probably seeing only a subset of the problem, hence overfitting on some data, hence converging much quicker, hence bad when restarting and presented other batches from the same database

I'm now going to implement another embedding-based loss more suitable for my problem.

**Comment by jmaye** [comment]
Happy to see this got the required attention: https://github.com/tensorflow/tensorflow/commit/f184d76cf0c927ba0f4f41bff22150f60befc844

Brief status, I can now train my network, reaching 80% of accuracy until now, I can stop and restart without change. I can also extract the embedding and use it with other images. Still need to evaluate the accuracy.

However, there still seems to be a bottleneck somewhere in terms of how the data is fed to the GPU. I can see that it regularly goes down to 0% usage although the CPU seems to run quite intensively. I will analyze further, maybe related to data augmentation (I have  a bunch more than in the example). In the C++ framework I used so far, this was more optimal.

Feel free to close the issue for now as I can get the results I want in terms of training. Hoping this was helpful to some people and I could share experience.

**Comment by aneergaard** [comment]
> What I discover:
> - probably for reproducibility, the random seed was set to 1 by default. Hence, when you use repeat on your dataset, the data samples are exactly in the same order. I changed that to None, working better.
> - with randomness set appropriately, the loss takes way more time to decrease
> - I can now stop learning and restore a checkpoint, I have exactly the same loss as when I excited and the same accuracy

Hi, just to clarify, what specifically did you do to change the seed? Was it in the tf.data.TextLineReader shuffle method, or somewhere else? I am running into the same problem where my training accuracy drastically decreases after each restart (ie. after each train/eval cycle using train_and_evaluate method) while the loss increases:
![image](https://user-images.githubusercontent.com/10213572/42647748-b61cf626-85b9-11e8-9ffb-ebb0b91ed09c.png)

Is this what you were experiencing?

**Comment by jmaye** [comment]
These problems were long time ago. Back then, I had to change the seed to None at start of the program. In newer versions of TF, they changed so that it does that automatically, but I agree all of that is quite tricky and not so well documented.
